src/alire/alire-origins-deployers-system-homebrew.adb

```diff
@@ -0,0 +1,273 @@
+         end Info_Callback;
+
+         Arr : JSON_Array;
+      begin
+         if Kind (Data) /= JSON_Array_Type then
+            raise JSON_Issue with "JSON info not JSON_Array";
+         end if;
+         Arr := Get (Data);
+         if Length (Arr) /= 1 then
+            raise JSON_Issue with "JSON info length /= 1";
+         end if;
+         if Kind (Get (Arr, 1)) /= JSON_Object_Type then
+            raise JSON_Issue with "JSON info (1) not JSON_Object";
+         end if;
+         Map_JSON_Object (Get (Arr, 1), Info_Callback'Access);
+      end;
+
+   --  exception
+   --     when E : JSON_Issue =>
+   --     Trace.Debug ("JSON issue: " & Ada.Exceptions.Exception_Message (E));
+   --        Available_Version := Null_Unbounded_String;
+   --        Installed_Version := Null_Unbounded_String;
+   end Get_Info;
+
+   -----------------------
+   -- Already_Installed --
+   -----------------------
+
+   overriding function Already_Installed (This : Deployer) return Boolean
+   is
+   begin
+      Trace.Debug ("already_installed? " & This.Base.Package_Name);
+      if Homebrew_Present then
+         declare
+            Cask_Name : constant String
+              := Homebrew_Prefix & "/Cellar/" & This.Base.Package_Name;
+            use Ada.Directories;
+         begin
+            return
+              Exists (Cask_Name)
+              and then Kind (Cask_Name) = Directory;
+         end;
+      else
+         Trace.Debug ("homebrew not found");
+         return False;
+      end if;
+   end Already_Installed;
+
+   ------------
+   -- Detect --
+   ------------
+
+   overriding
+   function Detect (This : Deployer) return Version_Outcomes.Outcome
+   is
+
+      function Get_Version_From_String (Candidate : String)
+                                       return Version_Outcomes.Outcome;
+      function Get_Version_From_String (Candidate : String)
+                                       return Version_Outcomes.Outcome
+      is
+         Regexp : constant String
+           := "(?:[[:digit:]]:)*([\d\.]+).*";
+         Matches : GNAT.Regpat.Match_Array (1 .. 1);
+         use type GNAT.Regpat.Match_Location;
+      begin
+         GNAT.Regpat.Match (Regexp, Candidate, Matches);
+         if Matches (1) /= GNAT.Regpat.No_Match then
+            Trace.Debug
+              ("Candidate version string: "
+                 & Candidate (Matches (1).First .. Matches (1).Last));
+            return
+              Version_Outcomes.New_Result
+                (Semantic_Versioning.Parse
+                   (Candidate (Matches (1).First .. Matches (1).Last),
+                    Relaxed => True));
+            --  Relaxed because ... not sure
+         else
+            return Version_Outcomes.Outcome_Failure
+              ("unexpected version format",
+               Report => False);
+         end if;
+      end Get_Version_From_String;
+
+   begin
+      Trace.Debug ("detect? " & This.Base.Package_Name);
+      if Homebrew_Present then
```

Comment:
Detection of the "distrib" should be done in the platform specific code. There's no need to do it here as far as I know.

Reply:
OK